coderd/prometheusmetrics/license.go

```diff
@@ -0,0 +1,125 @@
+package prometheusmetrics
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/codersdk"
+)
+
+type LicenseMetrics struct {
+	interval time.Duration
+	logger   slog.Logger
+	registry *prometheus.Registry
+
+	Entitlements atomic.Pointer[codersdk.Entitlements]
+}
+
+type LicenseMetricsOptions struct {
+	Interval time.Duration
+	Logger   slog.Logger
+	Registry *prometheus.Registry
+}
+
+func NewLicenseMetrics(opts *LicenseMetricsOptions) (*LicenseMetrics, error) {
+	if opts.Interval == 0 {
+		opts.Interval = 1 * time.Minute
+	}
+	if opts.Registry == nil {
+		opts.Registry = prometheus.NewRegistry()
+	}
+
+	return &LicenseMetrics{
+		interval: opts.Interval,
+		logger:   opts.Logger,
+		registry: opts.Registry,
+	}, nil
+}
+
+func (lm *LicenseMetrics) Collect(ctx context.Context) (func(), error) {
+	activeUsersGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "license",
+		Name:      "active_users",
+		Help:      `The number of active users.`,
+	}, []string{"entitled"}))
```

Comment:
What does the `entitled` label mean in this context? Same for `user_limit` below.

Reply:
The `entitled` label is defined in the Entitlements response, which means "enabled" if the "user_limit" is enforced. 

I left it there because you can't determine if the `active_users` is zero as nobody is using coderd at the moment or Entitlements entity returned zero. I know I can use a different query to fetch a number of active users (active API keys), but I would like to stick with Entitlements.

Please let me know your thoughts!

Reply:
Will this data only be available if a license is added? I could see a use case for monitoring active users even if you don't have a license (in this case, I could see all users being `entitled=false`).

Reply:
It doesn't make sense to me to have two different timeseries based on whether "user_limit" is enforced.  Since these are gauges, they will hold their values until changed, so if you switch from enforced to not enforced (or vice versa) the enforced timeseries holds the last value you wrote forever.

If we really need to expose in metrics whether it is enforced, it should be its own gauge.

Reply:
@spikecurtis The reason why I added the `entitled` label is to signal if the license is active. For instance, if you remove all licenses, entitlements structure is as follows:

```
"user_limit": {
   "entitlement": "not_entitled",
   "enabled": false,
   "actual": 1
 },
```

Can you share your idea of how we can model these metrics?

Reply:
FWIW in v1 we expose the following license-related metrics in case that helps:


```
coderd_license_expires_at_unix | gauge | Unix timestamp of the license expiry date.
coderd_license_issued_at_unix | gauge | Unix timestamp of the license issue date.
coderd_license_time_until_expires_days | gauge | Number of days until the license expires.
coderd_license_user_count | gauge | Number of active (non-dormant) users.
coderd_license_user_limit | gauge | Number of users allowed by the license.
```

Reply:
As we switched to the collector approach, I left only `coderd_license_active_users`, `coderd_license_limit_users`, and added `coderd_license_user_limit_enabled`.

I didn't include "expires at" metric, as it requires a DB call for licenses. We can't do this during `Collect()`.